### ISSUE TITLE
FP8 types should not participate in type promotion and should have no math ops defined on them

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#113663]
Right now, the following works
```
In [1]: import torch

In [2]: a=torch.randn(4, device="cuda")

In [3]: b=torch.zeros(4, device="cuda", dtype=torch.float8_e5m2)

In [4]: a+b
Out[4]: tensor([-0.6676, -0.6190, -0.8736,  0.1410], device='cuda:0')
```
But fp8 types are meant to work with per-tensor scales, and the above is a footgun that most people should avoid (it assumes the scales of the float8 tensor are 1), erroring out in this case and relying on operations implementation via subclassing fp8 tensor or some other means is the way to go. 
If someone indeed wants unscaled operations, they can explicitly cast fp8 tensor to a wider type, and rely on `torch.compile` to fuse this explicit cast away.  

cc @ezyang @gchanan @zou3519 @kadeng @yanbing-j @vkuzo @albanD

**Comment by albanD** [comment]
Should we consider removing all compute ops from fp8 (except the ones that also take scale) to avoid such issue?

**Comment by ngimel** [comment]
Yes! I haven't seen any compute ops implemented, thankfully, except those that would work due to type promotion.

**Comment by drisspg** [comment]
Yeah I think this makes sense

cc @malfet

**Comment by ezyang** [comment]
@ngimel wanna send us the patch? :)

**Comment by drisspg** [comment]
Should this restriction be applied to both "cpu" and "cuda" tensors / do we think that behavior like should also be disallowed.
https://github.com/pytorch/pytorch/blob/00b67193efc34702aaf70f2af158e0ce712c18b9/test/quantization/core/experimental/test_float8.py#L260-L271

Also for new all-reduce techniques (like the one mentioned here: https://arxiv.org/pdf/2310.18313.pdf) would it be feasible that we would want to implement reductions in float8.

**Comment by ngimel** [comment]
Reopening, as @drisspg points out in the comment above, `mul/eq/ne` should be disallowed for float8 types (added by #107586). I could be persuaded otherwise  about eq/ne, because it's a bits operation in some sense, but I'd rather we don't have it (`eq` could return true when scales are different). cc @malfet

**Comment by malfet** [comment]
Want to mention following usecase which will be broken in 2.2.0 by https://github.com/pytorch/pytorch/pull/113975: https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/14031
(Essentially f8 is used just as storage format, which makes inference slower but fit on some GPUs)

**Comment by jeffdaily** [comment]
@ngimel can you point me to the code from #107586 that added mul/eq/ne?  (My understanding of that PR was to mimic #104242 and only add 2 more F8 types, not add more features.)  I'm not disagreeing with this present issue, I'm just trying to learn so I don't inadvertently add support for something as I work on other F8-related PRs.

**Comment by KohakuBlueleaf** [comment]
> Want to mention following usecase which will be broken in 2.2.0 by #113975: [AUTOMATIC1111/stable-diffusion-webui#14031](https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/14031) (Essentially f8 is used just as storage format, which makes inference slower but fit on some GPUs)

I think autocast will not be broken by that? Seems like different problems 
And for gtx16/10series I use manualcast patch so not a problem.

**Comment by vkuzo** [comment]
> fp8 types are meant to work with per-tensor scales, and the above is a footgun that most people should avoid (it assumes the scales of the float8 tensor are 1)

+1 to removing any float8 ops from the aten API which assume anything about scales.  The way it's supposed to work in the current design is:
1. aten float8 dtypes and ops are unscaled
2. Python code implements the necessary scaling

**Comment by atalman** [comment]
Since this issue is still open As of Feb 14, 2024. Moving to release 2.2.2 milestone

**Comment by atalman** [comment]
removing this from release milestone.

**Comment by elismasilva** [comment]
if i load weights in fp8 and try to do training to fp8 i got errors like "addmm_cuda" not implemented for 'Float8_e4m3fn" so is not possible training and save weights in fp8 with pytorch?

**Comment by drisspg** [comment]
@elismasilva The fp8 dtype is intended to be use with a scale/scales for numerical accuracy. Most compute ops on tensors have been explicitly disabled since their API does not have a notion of this scale. The on exception being _scaled_mm which is still a private function. For an example of this dtype can be used you can take a look at https://github.com/pytorch-labs/float8_experimental

**Comment by vkuzo** [comment]
Reopening as we still have arithmetic on float8 ops defined: https://github.com/pytorch/pytorch/blob/0c81b398aba5d77ff9293f348e2ce28282ff7d54/c10/util/Float8_e4m3fn-inl.h#L31

Grabbing as we do want to fix this.

**Comment by ngimel** [comment]
Right, hopefully just deleting the arithmetic should be ok as dispatch should have been disabled.
